pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -32,122 +32,156 @@ import (
 
 const PluginClientTimeout = 45 * time.Second
 
-// draPluginClient encapsulates all dra plugin methods.
-type draPluginClient struct {
-	pluginName string
-	plugin     *Plugin
-}
-
 func NewDRAPluginClient(pluginName string) (drapb.NodeClient, error) {
 	if pluginName == "" {
 		return nil, fmt.Errorf("plugin name is empty")
 	}
 
-	existingPlugin := draPlugins.Get(pluginName)
+	existingPlugin := draPlugins.get(pluginName)
 	if existingPlugin == nil {
 		return nil, fmt.Errorf("plugin name %s not found in the list of registered DRA plugins", pluginName)
 	}
 
-	return &draPluginClient{
-		pluginName: pluginName,
-		plugin:     existingPlugin,
-	}, nil
+	return existingPlugin, nil
 }
 
-func (r *draPluginClient) NodePrepareResources(
+func (p *plugin) NodePrepareResources(
 	ctx context.Context,
 	req *drapb.NodePrepareResourcesRequest,
 	opts ...grpc.CallOption,
-) (resp *drapb.NodePrepareResourcesResponse, err error) {
+) (*drapb.NodePrepareResourcesResponse, error) {
 	logger := klog.FromContext(ctx)
 	logger.V(4).Info(log("calling NodePrepareResources rpc"), "request", req)
-	defer logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", resp, "err", err)
 
-	conn, err := r.plugin.getOrCreateGRPCConn()
+	conn, err := p.getOrCreateGRPCConn()
 	if err != nil {
 		return nil, err
 	}
-	nodeClient := drapb.NewNodeClient(conn)
-	nodeClientOld := drapbv1alpha2.NewNodeClient(conn)
 
 	ctx, cancel := context.WithTimeout(ctx, PluginClientTimeout)
 	defer cancel()
 
-	resp, err = nodeClient.NodePrepareResources(ctx, req)
+fallback:
+	p.RLock()
+	fallbackVersion := p.fallbackVersion
+	p.RUnlock()
+
+	if fallbackVersion == "v1alpha2" {
+		nodeClient := drapbv1alpha2.NewNodeClient(conn)
+		response := v1alpha2NodePrepareResouce(ctx, nodeClient, req.Claims)
+		logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", response, "err", nil)
+		return response, nil
+	}
+
+	nodeClient := drapb.NewNodeClient(conn)
+	response, err := nodeClient.NodePrepareResources(ctx, req)
 	if err != nil {
 		status, _ := grpcstatus.FromError(err)
 		if status.Code() == grpccodes.Unimplemented {
-			// Fall back to the older gRPC API.
-			resp = &drapb.NodePrepareResourcesResponse{
-				Claims: make(map[string]*drapb.NodePrepareResourceResponse),
-			}
-			err = nil
-			for _, claim := range req.Claims {
-				respOld, errOld := nodeClientOld.NodePrepareResource(ctx,
-					&drapbv1alpha2.NodePrepareResourceRequest{
-						Namespace:      claim.Namespace,
-						ClaimUid:       claim.Uid,
-						ClaimName:      claim.Name,
-						ResourceHandle: claim.ResourceHandle,
-					})
-				result := &drapb.NodePrepareResourceResponse{}
-				if errOld != nil {
-					result.Error = errOld.Error()
-				} else {
-					result.CDIDevices = respOld.CdiDevices
-				}
-				resp.Claims[claim.Uid] = result
-			}
+			p.Lock()
+			p.fallbackVersion = "v1alpha2"
```

Comment:
Looking at this code I'm worrying that handling new API versions would be difficult and would make the code even harder to understand. Can we come up with something more generic that would allow to add new handlers without complicating the processing flow?

Just an idea of the top of my head: keep supported versions in the list in the order they should be tried when the API is called first time. Remove unimplemented versions from the list if call returns `grpccodes.Unimplemented`.

Reply:
ok I'll give it a try